### PR TITLE
feature/issue-574

### PR DESCRIPTION
**Pull request by shes50103**
detect unexpected parsing error in parseDefMethodStatement() function.

### CODE REVIEW

**Comment by shes50103** [comment]
If there are any other parsing error which will crash parser please let me know

File: compiler/parser/statement_parsing.go [hunk 1]
```diff
@@ -51,6 +51,11 @@ func (p *Parser) parseDefMethodStatement() *ast.DefStatement {
 
 	p.nextToken()
 
+	if !p.curTokenIs(token.Ident) && !(p.peekTokenIs(token.Dot) && (p.curTokenIs(token.InstanceVariable) || p.curTokenIs(token.Constant) || p.curTokenIs(token.Self))) {
```
Comment (st0012) [inline compiler/parser/statement_parsing.go]: I think you should refactor this to like `!p.IsArgumentToken()`

**Comment by st0012** [comment]
@shes50103 I think we really need the recover mechanism, it's impossible to capture all invalid syntax cases.

**Comment by shes50103** [comment]
@st0012 
okay I will add the recover mechanism. I am thinking that if there is a panic inside parser without error detected, despite panic is recovered in parser but the error statement still make repl.go crash.
I still need to set the error mechanism in most of function to prevent error from edge case.

File: compiler/parser/statement_parsing.go (hunk 1 above)
Comment (shes50103) [reply compiler/parser/statement_parsing.go]: thanks, I will refactor it

File: compiler/parser/parser.go [hunk 2]
```diff
@@ -112,6 +112,13 @@ func New(l *lexer.Lexer) *Parser {
 
 // ParseProgram update program statements and return program
 func (p *Parser) ParseProgram() (*ast.Program, *errors.Error) {
+
+	defer func() {
+		if recover() != nil {
+			fmt.Println("Panic has been recover parser ")
```
Comment (st0012) [inline compiler/parser/parser.go]: This can be improved to `Parser recovered from panic`

Comment (shes50103) [reply compiler/parser/parser.go]: thanks!

**Pull request merged**